Workbench electron contextIsolation and nodeIntegration

davemfish opened the pull request:
This is a Workbench PR. It turns on `contextIsolation` and turns off `nodeIntegration` in the electron browser. Giving up node in the browser meant we have to bundle our browser code so it can be loaded without using `require`. So this PR also configures `vite` for bundling all the `src/renderer` code. 

To make sense of the new processes, maybe start by looking at the new `package.json` scripts. Development is easier now, just activate a python env with invest installed, and then `yarn start` from the workbench.

Lots of the other changes are there to setup the `contextBridge` in `preload`, exposing certain APIs to the renderer. To make sense of those, maybe start with `src/preload/preload.js`

Fixes #870 
fixes #864 
fixes #860 

## Checklist
- [ ] Updated HISTORY.rst (if these changes are user-facing)
- [ ] Updated the user's guide (if needed)
- [ ] Tested the affected models' UIs (if relevant)

davemfish commented on workbench/src/main/setupInvestHandlers.js:
```diff
@@ -103,8 +93,10 @@ export function setupInvestRunHandlers(investExe) {
-          event.reply(`invest-logging-${jobID}`, investLogfile);
+          const investLogfile = `${data}`.substring(
+            `${data}`.indexOf('[') + 1, `${data}`.indexOf(']')
+          );
```
This was to partially address #900

dcdenu4 commented on workbench/src/preload/api.js:
```diff
@@ -0,0 +1,66 @@
+  crypto: {
+    createHash: (algorithm) => crypto.createHash(algorithm)
+  },
+  electron: {
+    ipcRenderer: {
```
Do we have `electron` here as a nested level to help with maintaining understanding in the renderer code?

davemfish replied:
I figured this way it matches the actual `electron` API, where these methods are available at the `electron` namespace. Also it helps keep things organized right here since there's other non-electron methods on `Workbench`.

dcdenu4 commented on workbench/src/renderer/components/ResourcesLinks/index.jsx:
```diff
@@ -1,12 +1,14 @@
+
+const { ipcRenderer } = window.Workbench.electron;
+
+const UG_ROOT = 'http://releases.naturalcapitalproject.org/invest-userguide/latest';
```
Should this be `https`?

dcdenu4 commented on workbench/tests/renderer/app.test.js:
```diff
@@ -537,14 +535,41 @@ describe('InVEST subprocess testing', () => {
 
   afterEach(async () => {
-    mockInvestProc = null;
+    // mockInvestProc = null;
```
Want to leave this comment in?

davemfish replied:
heck no

dcdenu4 commented on workbench/src/main/main.js:
```diff
@@ -72,17 +78,17 @@ export const createWindow = async () => {
-      contextIsolation: false,
-      nodeIntegration: true,
-      preload: path.join(__dirname, '..', 'preload.js'),
+      contextIsolation: true,
+      nodeIntegration: false,
+      preload: path.join(__dirname, '../preload/preload.js'),
```
Alright!

dcdenu4 commented on workbench/src/main/menubar.js:
```diff
@@ -115,38 +119,44 @@ function openAboutWindow(parentWindow, isDevMode) {
-      contextIsolation: false,
-      enableRemoteModule: false,
-      nodeIntegration: true,
+      contextIsolation: true,
+      nodeIntegration: false,
```
I do like defining these, but I think this is the default for Electron now. In case we wanted to avoid writing these out in multiple places.

davemfish replied:
Yes you're right, these are the defaults now. Probably makes sense to remove them, at least once we've all noted the change.

davemfish replied:
I've removed these now.

dcdenu4 commented on src/natcap/invest/ui_server.py:
```diff
@@ -15,8 +16,15 @@
 
+PREFIX = 'api'
 app = Flask(__name__)
+CORS(app, resources={
+    f'/{PREFIX}/*': {
+        'origins': 'http://localhost:*'
+    }
```
This is allowing us to side step CORS security concerns and access files from the localhost?

davemfish replied:
Yes, it's specifically allowing CORS requests from localhost origins (vite dev server files) to endpoints matching `/{PREFIX}/*`

dcdenu4 commented on workbench/scripts/watch.js:
```diff
@@ -0,0 +1,20 @@
+const { createServer } = require('vite');
```
I might have missed it, but is this just a helpful tool for the bundling service?

davemfish replied:
This is the script that serves up the front-end code during development. It runs on `yarn run start`.

davemfish commented on workbench/src/renderer/app.jsx:
```diff
@@ -121,7 +120,9 @@ export default class App extends React.Component {
    */
   openInvestModel(job) {
-    const navID = crypto.randomBytes(16).toString('hex');
+    const navID = window.crypto.getRandomValues(
```
Ahh there's actually two unrelated crypto APIs at play. This one at `window.crypto` is a built-in browser API (nice!). 

The other one used elsewhere (to do some hashing in `InvestJob.js` I think) is the node module, which has a lot more functionality. I guess we had a use-case for that node module in the renderer, so that's why it's in preload. But right here in `app.jsx` it looks like the `window.crypto` was all we needed to get some random values.

dcdenu4 replied:
Thanks for clarifying that!

davemfish commented on workbench/tests/jest.overrides.js:
```diff
@@ -1,4 +1,27 @@
+
+  // jsdom does not implement these window APIs:
+  global.window.crypto = {
+    getRandomValues: () => [crypto.randomBytes(4).toString('hex')],
```
Yes good question, the dueling `crypto` api confusion continues...

Here we're setting up the APIs we need for the renderer tests, which run in a `jsdom` environment, which is _almost_ like the browser, but lacks some built-in browser APIs, like `window.crypto` and `window.fetch`. So we're patching in `crypto` into jsdom's `global.window`, but since the runtime for this setup script is node...well we only have the node module `crypto` available to generate some random data.

dcdenu4 replied:
Ahhh, got it, thanks.

davemfish commented on workbench/vite.config.js:
```diff
@@ -0,0 +1,31 @@
+export default defineConfig({
+  root: PROJECT_ROOT,
+  // base: empty string creates relative paths for assets.
+  // We need this because our files are embedded within an electron asar
```
https://www.electronjs.org/docs/latest/tutorial/application-distribution#with-an-app-source-code-archive

dcdenu4 replied:
Here I thought it was a typo! Thanks.

davemfish commented on workbench/electron-builder-config.js:
```diff
@@ -46,7 +46,7 @@ const config = {
   files: [
     'build/**/*',
-    'node_modules/**/*',
+    'node_modules/**/*', // TODO: probably should exclude any used by renderer.
```
Oh I forgot about this one. I think just to avoid duplication since some of these modules like react will have already been bundled by the vite build process. I'll take a stab at this now.

davemfish replied:
Okay I think I accomplished this. It's a little awkward, but electron-builder already knows to exclude anything listed in `package.json` `devDependencies`. So I moved all the front-end deps to that section. They are required for production, but vite already bundles them and puts them in `build/*` so they get packaged that way already. I added a readme section to explain this.

dcdenu4 replied:
Thanks Dave.

emlys commented on workbench/src/renderer/static/about.js:
```diff
@@ -0,0 +1,19 @@
+import { ipcMainChannels } from '../../main/ipcMainChannels';
+import { handleClick } from './handlers';
+
+const { ipcRenderer } = window.Workbench.electron;
```
Not sure if this is really a rule, but I associate `static/` with containing things that aren't code.

davemfish replied:
You're probably right. This was probably leftover from when this folder just held html documents. I truly don't know what the right naming conventions are though. Maybe this folder should be `src/menubar`? or `src/renderer/menubar`

davemfish replied:
I'm putting these .js files in `src/renderer/menubar/` and leaving the one .png here in `static`

emlys reviewed (commented):
@davemfish here's my initial thoughts and questions. Thanks for doing this huge update!

emlys commented on workbench/about.html:
```diff
@@ -79,7 +69,7 @@ <h2>Open-Source Licenses</h2>
           <td>
             <a
               href='http://numpy.org'
-              onClick="handleClick(event)">
```
Nevermind, I see below that the `onClick` is added later in a separate script.

davemfish commented on workbench/package.json:
```diff
@@ -30,37 +37,32 @@
   "license": "MIT",
   "dependencies": {
-    "bootstrap": "4.3.1",
+    "@babel/runtime": "^7.13.10",
```
This was supposed to be here instead of in `devDeps` all along. I think maybe things worked by coincidence because it was included by some other dependencies that were formerly in this section.

davemfish commented on workbench/src/renderer/components/ResourcesLinks/index.jsx:
```diff
@@ -38,7 +40,9 @@ const FORUM_TAGS = {
   event.preventDefault();
-  shell.openExternal(event.currentTarget.href);
+  ipcRenderer.send(
+    ipcMainChannels.OPEN_EXTERNAL_URL, event.currentTarget.href
```
Interesting, it sounds like sandboxing is even more strict than contextIsolation. But with this setup, we don't have access to `shell` in the renderer anymore. I'm not sure if there are vite things we could do to make it available without node, but `shell` operations really feel like they belong in `main` anyway.

So the options were either add the `shell` methods to the `contextBridge` in preload, or just use `shell` from the main process. I think it's a bit simpler overall to keep the contextBridge stuff as minimal as possible, just relying on the IPC that we already have setup.

davemfish commented on workbench/src/renderer/components/SettingsModal/SettingsStorage.js:
```diff
@@ -1,6 +1,6 @@
 import localforage from 'localforage';
 
-const logger = window.Workbench.getLogger(__filename.split('/').slice(-1)[0]);
+const logger = window.Workbench.getLogger('SettingsStorage.js');
```
Yes! it wasn't working properly before anyway, see #860

davemfish commented on workbench/tests/renderer/downloadmodal.test.js:
```diff
@@ -167,6 +166,9 @@ describe('Integration tests with main process', () => {
     };
+
+    // TODO: do we really want to mock electron.ipcRenderer.invoke
+    // or preload's api.electron.ipcRenderer.invoke?
```
Oh yes, thanks. This is tricky but I think it might be okay to remove this comment and leave as-is. We are mocking all parts of `electron` during all tests with `__mocks__/electron.js`. So when our application code calls `window.Workbench.electron.ipcRenderer.invoke` , that in turn calls `electron.ipcRenderer.invoke`. So if the latter is mocked, then that's slightly more "real" code being run during tests than if we mocked the former.

davemfish commented on src/natcap/invest/ui_server.py:
```diff
@@ -15,8 +16,15 @@
 LOGGER = logging.getLogger(__name__)
+logging.getLogger('flask_cors').level = logging.DEBUG
 
+PREFIX = 'api'
```
It just seems like a convention. I saw it everywhere when following the `flask-cors` examples. And it makes the pattern slightly narrower for the CORS exception.

davemfish commented on workbench/about.html:
```diff
@@ -2,50 +2,40 @@
-      <h3>InVEST Version: </h3> 
+    <span>
+      version: 
+      <em id='invest-version'></em>
```
Yeah we may want to do something else with these in the future. The problem is really that the window's menubar is completely outside the window's document (html). So making these pages with react & jsx isn't exactly the problem, they would still be in separate html documents unless we move them out of the menubar and into some other component in the main page.

emlys replied:
Ah okay. Eventually it could be worth moving them into a modal component or something so we have just one HTML page.

davemfish commented on workbench/src/renderer/components/InvestTab/index.jsx:
```diff
@@ -161,11 +160,6 @@ export default class InvestTab extends React.Component {
-    // here to be extra certain we avoid unexpected collisions
-    // of workspaceHash, which uniquely ids a job in the database
-    // in part by it's workspace directory.
-    args.workspace_dir = path.resolve(argsValues.workspace_dir);
```
No this is an oversight I think. I'll make an update. Thanks!

davemfish replied:
I updated this. The `path.resolve` now happens in the hashing function in `InvestJob`, right where it's needed. And `path.resolve` was added to the `contextBridge` to accomplish that.

davemfish commented:
@dcdenu4 and @emlys thanks for reviewing this so promptly. I think I addressed all your comments, but there's a new problem now that #918 is merged. It's solution is not compatible with `vite` builds and I actually don't think it works at all despite the builds passing on that PR. So I'll need to fix that, not sure if I'll do it here or in a separate PR.

davemfish commented on workbench/package.json:
```diff
@@ -5,14 +5,21 @@
-    "test": "jest --runInBand --testPathIgnorePatterns /tests/binary_tests/ /tests/sampledata_linkcheck/",
-    "test-electron-app": "jest --runInBand --testMatch **/tests/binary_tests/puppet.test.js",
-    "test-sampledata-registry": "jest --runInBand --testMatch **/tests/sampledata_linkcheck/sampledata_registry.test.js",
+    "test": "yarn run test-main && yarn run test-renderer && yarn run test-flask",
```
I thought the better of this and changed it back. It was a problem that there were multiple summary reports from jest instead of one big summary that is always printed to the console at the end.

emlys reviewed (approved):
Thanks @davemfish! I looked over the new changes and it makes sense to me.

dcdenu4 reviewed (commented):
Hey @davemfish thanks for your work on this and addressing my questions. Do we expect the tests to be passing? It looks they are all failing on the same thing. Just wanted to check before approving and merging.

davemfish commented:
> Hey @davemfish thanks for your work on this and addressing my questions. Do we expect the tests to be passing? It looks they are all failing on the same thing. Just wanted to check before approving and merging.

@dcdenu4 those tests have been failing for a while (#905 ) though now I see there's a new build failure before tests are even run. Maybe something to do with an major version upgrade of `twine`? I don't imagine it's related to this PR.

dcdenu4 reviewed (approved):
I don't think the failures are related to this PR either.
